examples/pipeline_wavernn/main.py

```diff
@@ -0,0 +1,420 @@
+
+        waveform = waveform.to(device)
+        specgram = specgram.to(device)
+        target = target.to(device)
+
+        output = model(waveform, specgram)
+        output, target = output.squeeze(1), target.squeeze(1)
+
+        if mode == "waveform":
+            output = output.transpose(1, 2)
+            target = target.long()
+
+        else:
+            # use mol mode
+            target = target.unsqueeze(-1)
+
+        loss = criterion(output, target)
+        loss_item = loss.item()
+        sums["loss"] += loss_item
+        metric("loss", loss_item)
+
+        optimizer.zero_grad()
+        loss.backward()
+
+        if args.clip_grad > 0:
+            gradient = torch.nn.utils.clip_grad_norm_(
+                model.parameters(), args.clip_grad
+            )
+            sums["gradient"] += gradient
+            metric("gradient", gradient.item())
+
+        optimizer.step()
+
+        metric("iteration", sums["iteration"])
+        metric("time", time() - start2)
+        metric.print()
+        sums["iteration"] += 1
+
+    avg_loss = sums["loss"] / len(data_loader)
+
+    metric = MetricLogger("train_epoch")
+    metric("epoch", epoch)
+    metric("loss", avg_loss)
+    metric("gradient", sums["gradient"] / len(data_loader))
+    metric("time", time() - start1)
+    metric.print()
+
+
+def validate(model, mode, criterion, data_loader, device, epoch):
+
+    with torch.no_grad():
+
+        model.eval()
+        sums = defaultdict(lambda: 0.0)
+        start = time()
+
+        for waveform, specgram, target in bg_iterator(data_loader, maxsize=2):
+
+            waveform = waveform.to(device)
+            specgram = specgram.to(device)
+            target = target.to(device)
+
+            output = model(waveform, specgram)
+            output, target = output.squeeze(1), target.squeeze(1)
+
+            if mode == "waveform":
+                output = output.transpose(1, 2)
+                target = target.long()
+
+            else:
+                # use mol mode
+                target = target.unsqueeze(-1)
+
+            loss = criterion(output, target)
+            sums["loss"] += loss.item()
+
+        avg_loss = sums["loss"] / len(data_loader)
+
+        metric = MetricLogger("validation")
+        metric("epoch", epoch)
+        metric("loss", avg_loss)
+        metric("time", time() - start)
+        metric.print()
+
+        return avg_loss
+
+
+def main(args):
+
+    devices = ["cuda" if torch.cuda.is_available() else "cpu"]
+
+    logging.info("Start time: {}".format(str(datetime.now())))
+
+    # Empty CUDA cache
+    torch.cuda.empty_cache()
+
+    melkwargs = {
+        "n_fft": args.n_fft,
+        "power": 1,
+        "hop_length": args.hop_length,
+        "win_length": args.win_length,
+    }
+
+    transforms = torch.nn.Sequential(
+        torchaudio.transforms.Spectrogram(**melkwargs),
+        # TODO Replace by torchaudio, once https://github.com/pytorch/audio/pull/593 is resolved
```

Comment:
can you move this comment to the definition of `linear_to_mel`?

Reply:
Yes, i moved it.